### ISSUE TITLE
Data inconsistency/errors when potentially reporting two pieces of data within the same millisecond

### ISSUE THREAD

**Issue by dpipemazo** [RedisTimeSeries/RedisTimeSeries#492]
This logically makes sense but I think brings some undefined behavior into the system. Same issue in theory (but more blatantly user error perhaps) if trying to write the same timestamp twice.

A bit of background: We're [adding Redis Time Series](https://github.com/elementary-robotics/atom/pull/359) as our default metrics engine for [Atom](http://atomdocs.io/), our microservice SDK built on redis streams + Docker. We chose to do this largely because of the slick integrations with Grafana -- the time series DB + integrations look great!

In writing unit tests it was found that if I made sequential calls to `TS.ADD` with timestamp of `*` I get the below error from redis (Redis 6.0.4 with redistimeseries 1.2.7 compiled from the github release)
```
redis.exceptions.ResponseError: Command # 1 (TS.ADD some_metric * 2020 RETENTION 60000) of pipeline caused error: TSDB: Timestamp cannot be older than the latest timestamp in the time series
```

Adding a sleep of 1ms or manually setting the timestamps to something unique works around the error as expected. 

I think this makes sense and don't see any real logical way around this at the moment but this is a change from the `XADD` behavior of streams where redis would auto-append a `-1` or similar to the end of the stream key if multiple keys had the same millisecond-level timestamp. Unfortunately this also can introduce timing-related, race-condition-esque bugs into production code as we were hoping to be able to use this to measure latencies in highly repetitive critical paths. It's unlikely we'd ever hit this error in production since the timing constraints are quite tight/unlikely, but it seems like something that maybe shouldn't throw an error that most people won't be planning for.

I do realize that the main difference here w.r.t. streams is that the timestamp IDs need to be integers and not strings in order to be useful for all of the aggregation. 

Interested to hear your thoughts on this behavior and if it is worthwhile to come up with a solution. Thanks for taking a look at this!

**Comment by danni-m** [comment]
Hi @dpipemazo, thanks for opening the issue.
Did you consider using your own timestamps (instead of `*`) in UT to avoid this race condition?

**Comment by dpipemazo** [comment]
@danni-m Yes, thank you. We are currently doing this as a workaround (using system wallclock), but will still see failures with this method without mitigation if we call the same code path twice within 1ms of each other. This may happen for example with a bunch of async jobs measuring latencies finishing at the same time.

**Comment by dpipemazo** [comment]
@danni-m I checked out `master` and can now see that the spirit of this bug is fixed, i.e. we won't get failures in this case but instead overwrites. This makes sense, but does add another failure mode into the mix: loss of data. Now, if I happen to have two calls to something I'm aggregating by COUNT or SUM that have the same timestamp (that occur within the same millisecond), I get overwrite/loss of the data in a failure mode that's difficult to detect.

I realize one workaround to this solution would be to use higher-fidelity timestamps on our side (microseconds would probably do), but I think this would then break out-of-the-box integrations with things like Grafana that have implicit dependence on the unix timestamps.

It would be nice to have a set of flags on TS.ADD that indicate what to do in the event of something with the same timestamp. Not necessarily the same as the `OOO` flag that seems to have not made the cut, but something along the lines of:

1. Fail if something with the same timestamp exists
2. Replace if something with the same timestamp exists
3. Aggregate if something with the same timestamp exists (SUM, MIN, MAX, COUNT, etc.)

This I would hope wouldn't be too difficult and would really help ensure there's no subtle data loss in systems that want to log things with <1ms fidelity.

**Comment by danni-m** [comment]
@dpipemazo Thanks for the feedback!

> Now, if I happen to have two calls to something I'm aggregating by COUNT or SUM that have the same timestamp (that occur within the same millisecond), I get overwrite/loss of the data in a failure mode that's difficult to detect.

Its important to note that you don't have to use `*`, you can use your own clock to determine the timestamp and avoid this issue.

> This may happen for example with a bunch of async jobs measuring latencies finishing at the same time.

This is where its up to you to decide how to handle multiple samples in the same timestamp, if you prefer you can use `TS.INCRBY` instead of `TS.ADD`. 

I just want to make sure we are solving an actual issue and not something that would happen in unittests only.

**Comment by dpipemazo** [comment]
> I just want to make sure we are solving an actual issue and not something that would happen in unittests only.

Absolutely agreed. I do think this is a real issue for us and I'll attempt to improve my explanation below.

> Its important to note that you don't have to use *, you can use your own clock to determine the timestamp and avoid this issue.

Yes, we've started doing this. I still have the same issue that if I make 2x calls within the same millisecond it will resolve to the same clock value.

> This is where its up to you to decide how to handle multiple samples in the same timestamp, if you prefer you can use TS.INCRBY instead of TS.ADD.

Thanks for pointing this out. I guess I can explain my use case a bit better. We're doing high-frequency processing of algorithms in a multi-thread/process environment on input data. I'm trying to track how long each call to my algorithm takes to complete. I have a few different scenarios:

--------------------------------------------------------------------------------------

Scenario A:
- Data comes in on a job queue to a worker with worker ID <worker_id>
- Obtain system timestamp for "start"
- Call algorithm(data)
- Obtain system timestamp for "end"
- Call `TS.ADD on key `<worker_id>:algorithmid:runtime` with timestamp equal to system clock and data equal to the delta between "start" and "end". Add `worker_id` and `algorithm_id` as labels s.t. we can use MRANGE to query for results across all workers, per-worker, etc.

In this scenario, if I have multiple calls to the algorithm that complete within the same millisecond I will lose data. I also have no real way of knowing this happened as this is now a silent failure. It's not unreasonable for simple algorithms for this to happen if two pieces of data happen to hit back to back.

-------------------------------------------------------------------------------

Scenario B:
- Data comes in on a job queue to a worker with worker ID <worker_id>
- Worker waits until N pieces of data are present in the queue before processing. This is fairly common for ML algorithms that run on GPUs as "batching" will result in overall greater throughput
- Note time item entered queue for "start"
- Call algorithm(list of data)
- Obtain system timestamp for "end"
- Call `TS.ADD on key `<worker_id>:algorithmid:runtime` with timestamp equal to system clock and data equal to the delta between the time each job hit the queue and "end". Add `worker_id` and `algorithm_id` as labels s.t. we can use MRANGE to query for results across all workers, per-worker, etc.

In the above scenario we will almost always silently drop data. I could use a different timestamp for the data, such as the time the item hit the queue, but again I may run into the same issue if multiple items hit the queue within the same millisecond which is not something I can really control.

-------------------------------------------------------------------------------------------------------------

Thanks for hearing a bit more about our use-case and scenarios. I do think it's ultimately something that could be worked around client-side but I also think this sort of feature will be quite nice for myself, my team, and perhaps other users moving forward. Generally I'd say it's difficult to make the assumption "two pieces of data will never hit the TSDB within the same millisecond" and currently this assumption is necessary in order to ensure no silent data loss.

**Comment by dpipemazo** [comment]
We actually have a third scenario that this will show up in which is interesting. The first two scenarios happen "on the edge" on our robots. When we're running on the edge we're typically connecting to redis over a shared `tmpfs` volume between our Docker containers and using a unix socket to minimize overhead. This is a very high-frequency data logging use case, perhaps atypically so for redis.

Our third scenario presents itself on the server side of our architecture, similarly related to the algorithm processing. We process too many algorithms per second on the edge to make calls to log the results to our server at such a high frequency. As such, we accept a small amount of latency and perform a batch creation of results in our server at a regular interval. We then expose a bulk creation endpoint on our server where it can take a list of algorithm result records to create and create individual rows in our relational database. 

We've found that making time-series queries across our relational database is not optimal/flexible enough in all use cases, so we are currently working on also logging metrics we want to track to Redis Time Series in order to serve the time-based queries from there. We can try to do some client-side aggregation before making calls to Redis Time Series but I'd prefer to add a bunch of individual data points to a single MADD call if possible. This is particularly relevant again for algorithm metadata such as timing -- we'd ideally like to track results on an individual basis. Even if we do all of the aggregation client-side in a single call from our bulk response endpoint there's no guarantee that two handlers for the endpoint server-side won't try to log results in the same millisecond window and thus drop/overwrite the results from one of the calls.

Finally, we were considering `*` to be a better/more accurate timestamp than using the compute timestamp due to potential drift of clocks across machines. I'm not sure what guarantees we get from our cloud providers due to how in-sync their wall clocks are, but if we're logging metrics from multiple servers to a single redis and are relying on the server-side clock for the timestamp I think it would make this particular bug even more difficult to trace/identify as if the clocks are even off by a second you'd see potential data loss if I made a call once per second to a server to log a metric and that call was issued to a different server behind a load balancer each time.

**Comment by dpipemazo** [comment]
Sorry for the multiple comments here but I think one final thought is worth mentioning:

Redis Time Series is **quite powerful** and we're using it in a way as a complete replacement for statsd/prometheus. Our code is planning to log to Redis Time Series directly. I think it makes sense that none of these features we're considering are useful when using something like a client-side aggregation tool such as statsd or prometheus, but our current understanding/belief is that through the use of redis pipelines and/or the bulk `MADD` query we no longer need to use these tools. 

Perhaps we shouldn't be doing this, but it would be preferable to us if so and so far it seems to be more or less working.

I think it helps to consider our use case in the above context since it would be a bit odd if a tool similar to prometheus or statsd had a requirement that no two calls to log a stat could occur within the same millisecond. Generally the timing level/granularity of the metric is less important and we really just want to make sure we're tracking 100% of the data.

**Comment by danni-m** [comment]
@dpipemazo Thanks! thats a lot of information to digest :)
Did you guys use statsd to aggregate your input before using RedisTimeSeries?

**Comment by dpipemazo** [comment]
@danni-m Yes, sorry about the length of the post(s) -- was mainly hoping to convey as much context as possible.

> Did you guys use statsd to aggregate your input before using RedisTimeSeries?

We've been struggling to find a good set of tools for this at Elementary. A few years ago (at another role, developing edge IoT firmware) I used statsd client-side and then pushed to a cloud-hosted Timescale DB instance. We logged all of our data to statsd over a local UDP socket and then wrote a custom statsd sink that ran periodically to do the downsampling/aggregation and forward along the summary stats to Timescale.

We looked into prometheus and found it quite nice, but generally lacking all of the language support/flexibility we needed on the client side and also adding quite a bit of complexity/overhead to the system. The pull model of collecting stats (compared to statsd push-model) makes sense in a lot of use cases but was also generally worse for us compared to a push-type model. A good amount of the downsides of the push-type model are negated through (1) being on the same compute as the redis most of the time and (2) using a unix socket for data transfer. We'd also then need to expose a stats endpoint on each of our microservices which was also not ideal.

I then started using Redis Time Series and was quite impressed. Our [SDK](https://github.com/elementary-robotics/atom) already tightly integrates with redis (we move nearly 100% of our data on the edge through redis streams) and we have a [WIP PR](https://github.com/elementary-robotics/atom/pull/359) where the integration with Redis Time Series can be seen. We can get quite good instrumentation for free on the edge through using Redis Time Series + periodic RDB backups + grafana for visualization. Cloud-side for now we can just back up the RDB files periodically for point-in-time snapshots which is good enough and very low cost. 

We could definitely change the SDK to log to StatsD and then write a sink which forwards along to Redis Time Series as the backend and get more or less the same outcome as what I'm asking for above. It would mean needing to run another container/process on the edge and seems potentially avoidable. It would also mean needing to write the StatsD sink if it's not already available.

**Comment by danni-m** [comment]
@dpipemazo Just an idea, what if you would report to a stream key that would then be automatically aggregated to 4 keys (count, min, max, avg).
This might be doable via https://github.com/RedisGears/RedisGears

**Comment by dpipemazo** [comment]
@danni-m Sure, we could look into that. Thanks for the tip.

Before investing some time in an external solution on our end (gears, StatsD), are you essentially saying this is not something you think it's Redis Time Series's job to handle and this is going to be accepted as its behavior moving forward? I think you're really close to not needing any additional external solution through something like the flags I proposed and I'd be happy to take a look at implementing the feature and submitting a PR if it's something you're open to.

**Comment by ashtul** [comment]
@dpipemazo 
Thank you for the input and insights. 
This is a very interesting idea and I can see the usefulness of it. 
To accommodate your third suggestion, the series will have to hold the data differently since we'll have to save two values (value,count) since we want real average and not just averaging two numbers at the time. 
Since this would require some major changes to the code, it is hard to tell if and when would we implement it.

**Comment by dpipemazo** [comment]
> To accommodate your third suggestion, the series will have to hold the data differently since we'll have to save two values (value,count) since we want real average and not just averaging two numbers at the time.

Yes, I agree about this, though I think we could perhaps get most of the way there quickly? For a decent amount of use cases (min, max, sum, count) would we be able to get away without having to change the data structure?

**Comment by danni-m** [comment]
@dpipemazo While I agree we are not far from having "StatsD like functionality", implementing one should be a thoughtful process and will take time.
Working with Redis gears will allow us both to "mock" this behaviour and have better understanding of what is needed.

Im not sure we want to mix "database" like functionality with something that resembles more of a "stream processing/metrics aggregation", it might be in the same module, but not necessarily work with the same command as inserting/updating a single sample.

Does that make sense?

**Comment by ashtul** [comment]
@dpipemazo
Adding the option to aggregate data probably will not happen in this release as it will require a new type of chunk which will take time to develop. 

You could do it by using TS.INCRBY as @danni-m suggested. You could even have average by having a second series where you incrby by 1 and then do the decision on the client. 

We debate adding a flag that will change the behavior to the old one where an out of order sample will be rejected and return an error. Will take solve the issue the you see?

**Comment by dpipemazo** [comment]
> Adding the option to aggregate data probably will not happen in this release as it will require a new type of chunk which will take time to develop.

I'm still not sure I follow this. If we limit the aggregation options to `SUM`, `MIN`, `MAX` there shouldn't be additional chunk work required. Since the current behavior is just "replace", instead of doing that you can either do a sum, min or max and move on. This will solve the data consistency issues I've been discussing and while I agree it's not "fully featured" (can't do average, P95, etc.) I think holding up the work on that would be a mistake compared to the value it would bring.

**Comment by ashtul** [comment]
@dpipemazo 
You can get `SUM` by using `TS.INCRBY` but you are right that some other metrics such as `MIN`, `MAX' and maybe also `FIRST`, `LAST` can be done with the current chunks.
I will take it to the team but regardless, it probably won't happen before the next release.

**Comment by dpipemazo** [comment]
@ashtul Thank you. I **really** appreciate you taking it to the team and understand the timing/release cycle. I'm quite excited to see this feature move forward.

**Comment by dpipemazo** [comment]
Closing this as the `SUM` duplicate sample policy was merged in with #565 . Thank you to @danni-m and the RTS team for working this in! It will help us a lot.
